nbconvert/filters/ansi.py

```diff
@@ -1,155 +1,264 @@
     'ansi2latex'
 ]
 
-ansi_colormap = {
-    '30': 'ansiblack',
-    '31': 'ansired',
-    '32': 'ansigreen',
-    '33': 'ansiyellow',
-    '34': 'ansiblue',
-    '35': 'ansipurple',
-    '36': 'ansicyan',
-    '37': 'ansigrey',
-    '01': 'ansibold',
-}
-
-html_escapes = {
-    '<': '&lt;',
-    '>': '&gt;',
-    "'": '&apos;',
-    '"': '&quot;',
-    '`': '&#96;',
-}
-ansi_re = re.compile('\x1b' + r'\[([\dA-Fa-f;]*?)m')
+_FG_HTML = (
+    'ansiblack',
+    'ansired',
+    'ansigreen',
+    'ansiyellow',
+    'ansiblue',
+    'ansipurple',
+    'ansicyan',
+    'ansigray',
+)
+
+_BG_HTML = (
+    'ansibgblack',
+    'ansibgred',
+    'ansibggreen',
+    'ansibgyellow',
+    'ansibgblue',
+    'ansibgpurple',
+    'ansibgcyan',
+    'ansibggray',
+)
+
+_FG_LATEX = (
+    'black',
+    'red',
+    'green',
+    'brown',
+    'blue',
+    'purple',
+    'cyan',
+    'lightgray',
+)
+
+_BG_LATEX = (
+    'darkgray',
+    'lightred',
+    'lightgreen',
+    'yellow',
+    'lightblue',
+    'lightpurple',
+    'lightcyan',
+    'white',
+)
+
 
 def ansi2html(text):
     """
-    Convert ansi colors to html colors.
-    
+    Convert ANSI colors to HTML colors.
+
     Parameters
     ----------
     text : str
-        Text containing ansi colors to convert to html
+        Text containing ANSI colors to convert to HTML
+
     """
+    text = str(jinja2.utils.escape(text))
+    return _ansi2anything(text, _htmlconverter)
 
-    # do ampersand first
-    text = text.replace('&', '&amp;')
-    
-    for c, escape in html_escapes.items():
-        text = text.replace(c, escape)
-
-    m = ansi_re.search(text)
-    opened = False
-    cmds = []
-    opener = ''
-    closer = ''
-    while m:
-        cmds = m.groups()[0].split(';')
-        closer = '</span>' if opened else ''
-        
-        # True if there is there more than one element in cmds, *or*
-        # if there is only one but it is not equal to a string of zeroes.
-        opened = len(cmds) > 1 or cmds[0] != '0' * len(cmds[0])
-        classes = []
-        for cmd in cmds:
-            if cmd in ansi_colormap:
-                classes.append(ansi_colormap[cmd])
-
-        if classes:
-            opener = '<span class="%s">' % (' '.join(classes))
-        else:
-            opener = ''
-        text = re.sub(ansi_re, closer + opener, text, 1)
 
-        m = ansi_re.search(text)
+def ansi2latex(text):
+    """
+    Convert ANSI colors to LaTeX colors.
+
+    Parameters
+    ----------
+    text : str
+        Text containing ANSI colors to convert to LaTeX
+
+    """
+    return _ansi2anything(text, _latexconverter)
+
+
+def _htmlconverter(fg, bg, bold):
+    """
+    Return start and end tags for given foreground/background/bold.
+
+    """
+    if (fg, bg, bold) == (None, None, False):
+        return '', ''
+
+    classes = []
+    styles = []
 
-    if opened:
-        text += '</span>'
-    return text
+    if isinstance(fg, int):
+        classes.append(_FG_HTML[fg])
```

Comment:
Well, "impossible" is a very strong word, but yes, if `fg` (or `bg`) is a single `int`, it is supposed to be within `range(8)`. AFAICT, there is no code path that allows otherwise (assuming no monkey-patching and stuff).